lua/bson_lua_wrapper.c

```diff
@@ -0,0 +1,496 @@
+
+      //Assume array if no key is provided
+      if (key == NULL) {
+        bson_array_add_double((BsonArray *)bson, (double)value);
+      }
+      else {
+        bson_object_put_double((BsonObject *)bson, key, (double)value);
+      }
+      break;
+    }
+    case TYPE_BOOLEAN: {
+      if (lua_isboolean(L, -1) == 0) {
+        snprintf(errorMessage, 255, "Expected boolean, but value was of type: %s", 
+             lua_typename(L, lua_type(L, -1)));
+        return -1;
+      }
+      int value = lua_toboolean(L, -1);
+
+      //Assume array if no key is provided
+      if (key == NULL) {
+        bson_array_add_bool((BsonArray *)bson, (bson_boolean)value);
+      }
+      else {
+        bson_object_put_bool((BsonObject *)bson, key, (bson_boolean)value);
+      }
+      break;
+    }
+    default: {
+      snprintf(errorMessage, 255, "Error parsing table, unrecognized type: %i", (int)type);
+      return -1;
+    }
+  }
+  lua_pop(L, 1); //Stack: [{type, value}]
+  return 0;
+}
+
+/*
+  @brief Convert a BSON-formatted byte string into a Lua table.
+    Lua arguments:
+      1 (string) - BSON-formatted byte string to be converted
+    Lua return values:
+      1 (table) - Lua table parsed from the given BSON string.
+        Entries in this map are provided in the following format:
+          map[key] = {
+            type = element_type,
+            value = val
+          }
+
+  @param L - Lua state, used to modify the Lua stack
+
+  @return The number of return values pushed to the stack
+*/
+static int map_from_bytes(lua_State *L) {
+  if (lua_isstring(L, 1) == 0) {
+    return luaL_argerror(L, 1, "Arg 1 must be a Lua String");
+  }
+  size_t len;
+  const char *byte_string = lua_tolstring(L, 1, &len);
+
+  uint8_t *bytes = malloc(len);
+  
+  //Convert lua string to byte array to pass to the BSON library
+  size_t index = 0;
+  for (index = 0; index < len; index++) {
+    bytes[index] = byte_string[index] & 0xFF;
+  }
+
+  BsonObject obj = bson_object_from_bytes(bytes);
+  free(bytes);
+  
+  char errorMessage[255];
+  const int result = bson_object_to_table(L, &obj, errorMessage);
+  bson_object_deinitialize(&obj);
+
+  if (result != 0) {
+    return luaL_argerror(L, 1, errorMessage);
+  }
+  return 1;
+}
+
+/*
+  @brief Convert a BSON object into a Lua table.
+    Lua return values:
+      1 (table) - Lua table generated from the given BSON object.
+        Entries in this map are provided in the following format:
+          key = {
+            type = element_type,
+            value = val
+          }
+
+  @param L - Lua state, used to modify the Lua stack
+  @param obj - The BSON object to be converted
+  @param errorMessage - Populated in the case of an error
+
+  @return 0 on success, non-zero on failure
+*/
+static int bson_object_to_table(lua_State *L, BsonObject *obj, char *errorMessage) {
+  const MapIterator iter = bson_object_iterator(obj);
+  BsonObjectEntry entry = bson_object_iterator_next(&iter);
```

Comment:
@jacobkeeler It is compilation warning here:
" warning: passing argument 1 of ‘bson_object_iterator_next’ discards ‘const’ qualifier from pointer target type"
Please take a look.

The same warning in line 376